### PR TITLE
v2 update of metric schema

### PR DESCRIPTION
**Pull request by woodsaj**
- remove the "Metric" field from MetricData and MetricDefinition as it
  is redundant.  All clients currently set both Name and Metric fields to
  the same value.  Metrictank index only uses the Name field
- Change the Id field to be a [16]byte array instead of a string. The Id
  will just be the md5 hash of the other fields. the orgId prefix has been
  removed from the id and instead the orgId field is used in the hash.

### CODE REVIEW

**Comment by DanCech** [comment]
I'm confused by the Id change, it seems like the current format would be more conducive to having the client calculate and send a hashed Id instead of the full metric definition, since it could just send the md5 sum portion and we can assemble the full Id by adding org id to the hashed portion. That said, the API keys do contain the org id, so it shouldn't be that difficult for crng to calculate the full hash either.

File: metric.go [hunk 1]
```diff
@@ -32,12 +33,28 @@ type PartitionedMetric interface {
...
+	return fmt.Sprintf("%x", m)
+}
+
+func (m MetricId) MarshalJSON() ([]byte, error) {
+	return []byte('"' + m.String() + '"'), nil
+}
+
+func (m *MetricId) UnmarshalJSON(data []byte) error {
+	dst := make([]byte, 16)
+	_, err := hex.Decode(dst, data[1:len(data)-1])
```
Comment (Dieterbe) [inline metric.go]: We should probably be more defensive here

**Review by Dieterbe — CHANGES_REQUESTED** [review]
I Need to think more about this. String -> byte array is great and long overdue. Migration will be interesting. Shall we just create new series and rely on MT merging them ?

**Comment by woodsaj** [comment]
@DanCech that is a good point.  I think we should move the orgId out of the id hash.  My motivation for putting it in there was to ensure the ids would be unique across customers in a multitenant environment.  However, i think we can leave that up to the consumers of the metricDefs for how they accomplish that. eg. the inMemory index already creates per-org indexes.  And the cassandra-idx can continue using the same format it currently does by using `sprintf("%d.%x", m.OrgId, m.Id)` or it can change the table format to use a composite key (orgId, partition, id,).   A similar approach can be used for the cassandra_store.  And future stores (BigTable, CosmosDB) can use their own approaches.

**Comment by woodsaj** [comment]
@Dieterbe if we dont include the OrgId when generating the Hash, then i think the keys will stay the same.  If the Id were to change it would result in indexs doubling, at least in the number of metricDefs.  However the tree index and the TagIndex are based on the Name and Tags, which would be the same with new and old metricDefs, so those indexes would not grow.

We should not need to migrate immediately, as cassandra-idx and cassandra-store can just be updated to use `fmt.Sprintf("%d.%x", m.OrgId, m.Id)`   instead of just `m.Id`

File: metric.go [hunk 2]
```diff
@@ -32,12 +33,28 @@ type PartitionedMetric interface {
 
 //go:generate msgp
 
+type MetricId [16]byte
```
Comment (replay) [inline metric.go]: Wouldn't it make sense to make this type identical with the `MetricID` here and then just embed it into `MetricData`: https://github.com/grafana/metrictank/blob/master/idx/idx.go#L38

Then we could directly use it as it is when ingesting into the in-memory index.

Comment (replay) [reply metric.go]: IIRC our long term plan was to replace the keys in `MemoryIdx.DefById` with a numeric type, right? If this `schema.MetricId` would have the org id like `idx.MetricID` does, then we could use this type as key for `DefById` in the future.

**Comment by Dieterbe** [comment]
> Wouldn't it make sense to make this type identical with the MetricID here and then just embed it into MetricData: 

IMHO no because the structure already has an org id field. Why specify it twice especially when it's not needed ? I like the philosophy of singletenant IDs (not including orgids) and where it needs to be org aware (eg MT index), leave that up to the implementation

**Comment by Dieterbe** [comment]
> IIRC our long term plan was to replace the keys in MemoryIdx.DefById with a numeric type, right?

For the record I believe a 16B Array and 2x64 uints are equivalent in terms of memory usage and GC overhead (none) so whichever is easier to work with should be preferred. It seems to be the former

**Comment by replay** [comment]
@Dieterbe i did not mean to duplicate the ID. if `schema.MetricId` contains the org id and `MetricId` is embedded in `MetricData` then we can remove `orgId` from `MetricData`

File: metric.go [hunk 3]
```diff
@@ -56,9 +73,6 @@ func (m *MetricData) Validate() error {
 	if m.Name == "" {
 		return errInvalidEmptyName
 	}
-	if m.Metric == "" {
-		return errInvalidEmptyMetric
```
Comment (Dieterbe) [inline metric.go]: remove the errInvalidEmptyMetric initialisation also

File: metric.go (hunk 1 above)
Comment (Dieterbe) [reply metric.go]: (meaning return the err if it's non nil)

**Comment by Dieterbe** [comment]
for anyone interested in the history behind some of these fields, https://github.com/grafana/metrictank/issues/30 takes you through some early design convo's.

Note that tsdbgw sets - in the metrics2.0 ingestion receiver - the Metric property to Name, if it was previously `""`, so still might be non-equal based on what the customer sent. which raises the question, amongst the senders we know of, are .Metric and .Name set to the same value? I verified the following:

Y carbon-relay-ng's grafanaNet and kafkaMdm route (via its parseMetric function)
Y tsdb-gw's carbon input (via its parseMetric function)
Y usage metrics generated by tsdb-usage (note: MT no longer generates usage metrics)
Y https://github.com/grafana/hosted-metrics-sender-example go and python example
Y MT carbon input
N an [early version of the go example sender](https://gist.github.com/Dieterbe/b7e3addb9ead628571a677ef490afbb1)


So practically for the vast majority the resulting id will remain the same, and in a few cases (customers using their own senders, eg based on the original gist) we will create new series, but MT would merge them. we can also advise customers to upgrade custom senders to the new schema.  or we could rewrite old Metric values to Name values of course, which raises the question of how we would migrate. it's one thing to update MT (and things would be equivalent); it's another to deploy an updated MT which uses the v2 schema, we still need to support transitions where we upgrade tsbgw and may have mixed data in kafka, and MT needs to support reading it all in. So MT needs to support both schema's. this is where things get interesting.

@replay re the orgid. that seems ok to me. But both your suggestion and the current approach seem very similar to me and I don't see a clear reason to prefer one over the other.
with your suggestion, we could save a few lines in MT because we can reuse the same type, and it seems nice to have 1 standard location to define what an id (incorporating the orgid) would look like, but the downside is, everywhere we want to use the id that doesn't incorporate the orgid, we need to access a sub-property which is a bit less intuitive and a bit more to type. Is there any stronger reason to prefer things one way or another?

**Comment by replay** [comment]
@dieterbe nah, no better reason. it just seemed nice if on  ingest we could directly take the `schema.MetricId` property and look it up on the `DefByIds` index in order to get the according MD, without any copying or constructing the object first.

**Pull request closed without merging**